src/main/java/com/google/cloud/tools/appengine/AppEngineDescriptor.java

```diff
@@ -99,24 +106,90 @@ public boolean isJava8() {
+   *   <env-var name="key" value="val" />
+   * </env-variables>
+   * }
+   * </pre>
+   * This will construct a map of the form {[key, val], ...}
+   *
+   * @return a map representing the environment variable settings in the appengine-web.xml.
+   */
+  public Map<String, String> getEnvironment() {
+    Node environmentParentNode = getNode(document, "appengine-web-app", "env-variables");
+    return getAttributeMap(environmentParentNode, "env-var", "name", "value");
+  }
+
+  private static String getText(Node node) {
+    if (node != null) {
+      try {
+        return node.getTextContent();
+      } catch (DOMException ex) {
+        throw new AppEngineException("Failed to parse text content from node "
+            + node.getNodeName());
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * Returns a map formed from the attributes of the nodes contained within the parent node.
+   */
+  private static Map<String, String> getAttributeMap(Node parent,
+                                                     String nodeName,
+                                                     String keyAttributeName,
+                                                     String valueAttributeName) {
+    if (parent != null) {
+      Map<String, String> nameValueAttributeMap = Maps.newHashMap();
+
+      if (parent.hasChildNodes()) {
+        for (int i = 0; i < parent.getChildNodes().getLength(); i++) {
+          Node child = parent.getChildNodes().item(i);
+          NamedNodeMap attributeMap = child.getAttributes();
+
+          if (nodeName.equals(child.getNodeName()) && attributeMap != null) {
+            Node keyNode = attributeMap.getNamedItem(keyAttributeName);
+
+            if (keyNode != null) {
+              Node valueNode = attributeMap.getNamedItem(valueAttributeName);
+              try {
+                nameValueAttributeMap.put(keyNode.getTextContent(), valueNode.getTextContent());
+              } catch (DOMException ex) {
+                throw new AppEngineException("Failed to parse value from attribute node "
+                    + keyNode.getNodeName());
+              }
             }
           }
         }
       }
-      return null;
-    } catch (DOMException ex) {
-      // this shouldn't happen barring a very funky DOM implementation
-      return null;
+
+      return nameValueAttributeMap;
+    }
+
+    return null;
+  }
+
+  /**
+   * Returns the first node found matching the given name contained within the parent node.
+   */
+  private static Node getNode(Document doc, String parentNodeName, String targetNodeName) {
+    NodeList parentElements = doc.getElementsByTagNameNS(APP_ENGINE_NAMESPACE, parentNodeName);
+    if (parentElements.getLength() > 0) {
+      Node parent = parentElements.item(0);
+      if (parent.hasChildNodes()) {
+        for (int i = 0; i < parent.getChildNodes().getLength(); i++) {
+          Node child = parent.getChildNodes().item(i);
+          if (child.getNodeName().equals(targetNodeName)) {
```

Comment:
could you be more specific regarding your last sentence? e.g. what could go wrong and do you think I should be handling it here?